lib/services/aiSynthesis/buffer-reduce.subject.ts

```
@@ -0,0 +1,37 @@
+import { Subject } from 'rxjs';
+
+export type Predicate<T> = (buffer: T) => PredicateResult;
+export type PredicateResult = boolean | 'stop';
+export type Accumulator<T> = (buffer: T, value: T) => T;
+
+export class BufferedReducerSubject<T> extends Subject<T> {
+  private buffer?: T;
+
+  private buffering = true;
+
+  constructor(private readonly predicate: Predicate<T>, private readonly accumulator: Accumulator<T>) {
+    super();
+  }
+
+  next(value: T): void {
+    if (!this.buffering) {
+      super.next(value);
+      return;
+    }
+
+    this.buffer = this.buffer ? this.accumulator(this.buffer, value) : value;
+
+    const result = this.predicate(this.buffer);
+    if (result === 'stop') {
```

its a bit weird that the predicate returns true/false/stop.

would it make sense instead to just return true/false and do a try/catch around it to get the 'stop' usecase?

then the code that calls the buffer would throw the error instead of returning 'stop'. also I guess the error message could be better than just the class name